snap/squashfs/squashfs.go

```diff
@@ -301,8 +304,101 @@ func (s *Snap) ListDir(dirPath string) ([]string, error) {
 	return directoryContents, nil
 }
 
+const maxErrPaths = 10
+
+type errPathsNotReadable struct {
+	paths []string
+}
+
+func (e *errPathsNotReadable) accumulate(p string, fi os.FileInfo) error {
+	if len(e.paths) >= maxErrPaths {
+		return e
+	}
+	if st, ok := fi.Sys().(*syscall.Stat_t); ok {
+		e.paths = append(e.paths, fmt.Sprintf("%s (owner %v:%v mode %#03o)", p, st.Uid, st.Gid, fi.Mode().Perm()))
+	} else {
+		e.paths = append(e.paths, p)
+	}
+	return nil
+}
+
+func (e *errPathsNotReadable) asErr() error {
+	if len(e.paths) > 0 {
+		return e
+	}
+	return nil
+}
+
+func (e *errPathsNotReadable) Error() string {
+	var b bytes.Buffer
+
+	b.WriteString("cannot access the following locations in the snap source directory:\n")
+	for _, p := range e.paths {
+		fmt.Fprintf(&b, "- ")
+		fmt.Fprintf(&b, p)
+		fmt.Fprintf(&b, "\n")
+	}
+	if len(e.paths) == maxErrPaths {
+		fmt.Fprintf(&b, "- too many errors, listing first %v entries\n", maxErrPaths)
+	}
+	return b.String()
+}
+
+// verifyContentAccessibleForBuild checks whether the content under source
+// directory is usable to the user and can be represented by mksquashfs.
+func verifyContentAccessibleForBuild(sourceDir string) error {
+	var errPaths errPathsNotReadable
+
+	withSlash := filepath.Clean(sourceDir) + "/"
+	err := filepath.Walk(withSlash, func(path string, st os.FileInfo, err error) error {
+		if err != nil {
+			if !os.IsPermission(err) {
+				return err
+			}
+			// accumulate permission errors
+			return errPaths.accumulate(strings.TrimPrefix(path, withSlash), st)
+		}
+		mode := st.Mode()
+		if !mode.IsRegular() && !mode.IsDir() {
+			// device nodes are just recreated by mksquashfs
+			return nil
+		}
+		if mode.IsRegular() && st.Size() == 0 {
+			// empty files are also recreated
+			return nil
+		}
+
+		f, err := os.Open(path)
+		if err != nil {
+			if !os.IsPermission(err) {
+				return err
+			}
+			// accumulate permission errors
+			if err = errPaths.accumulate(strings.TrimPrefix(path, withSlash), st); err != nil {
```

Comment:
We leak a fd here, don't we? "f" won't get closed or am I missing something? If we extract this into a helper we could use defer f.close()

Reply:
That's an error branch, `err != nil` and `f == nil` at this point.

Reply:
Indeed, silly me. Ignore that.